embedded-hal/src/i2c.rs

```diff
@@ -14,128 +14,140 @@
-//!         // ...
-//! #       Ok(())
-//!     }
-//!     fn write_iter_read<B: IntoIterator<Item = u8>>(&mut self, addr: u8, bytes: B, buffer: &mut [u8]) -> Result<(), Self::Error> {
-//!         // ...
-//! #       Ok(())
-//!     }
-//!     fn transaction<'a>(&mut self, address: u8, operations: &mut [Operation<'a>]) -> Result<(), Self::Error> {
-//!         // ...
-//! #       Ok(())
-//!     }
-//!     fn transaction_iter<'a, O: IntoIterator<Item = Operation<'a>>>(&mut self, address: u8, operations: O) -> Result<(), Self::Error> {
-//!         // ...
-//! #       Ok(())
-//!     }
-//! }
+#![doc = include_str!("i2c-shared-bus.svg")]
 //!
-//! impl I2c<TenBitAddress> for I2c0
-//! {
-//!     fn read(&mut self, addr: u16, buffer: &mut [u8]) -> Result<(), Self::Error> {
-//!         // ...
-//! #       Ok(())
-//!     }
-//!     fn write(&mut self, addr: u16, bytes: &[u8]) -> Result<(), Self::Error> {
-//!         // ...
-//! #       Ok(())
-//!     }
-//!     fn write_iter<B: IntoIterator<Item = u8>>(&mut self, addr: u16, bytes: B) -> Result<(), Self::Error> {
-//!         // ...
-//! #       Ok(())
-//!     }
-//!     fn write_read(&mut self, addr: u16, bytes: &[u8], buffer: &mut [u8]) -> Result<(), Self::Error> {
-//!         // ...
-//! #       Ok(())
-//!     }
-//!     fn write_iter_read<B: IntoIterator<Item = u8>>(&mut self, addr: u16, bytes: B, buffer: &mut [u8]) -> Result<(), Self::Error> {
-//!         // ...
-//! #       Ok(())
-//!     }
-//!     fn transaction<'a>(&mut self, address: u16, operations: &mut [Operation<'a>]) -> Result<(), Self::Error> {
-//!         // ...
-//! #       Ok(())
-//!     }
-//!     fn transaction_iter<'a, O: IntoIterator<Item = Operation<'a>>>(&mut self, address: u16, operations: O) -> Result<(), Self::Error> {
-//!         // ...
-//! #       Ok(())
-//!     }
-//! }
-//! ```
+//! This bus sharing is common when having multiple I2C devices in the same board, since it uses fewer MCU
+//! pins (`2` instead of `2*n`), and fewer MCU I2C peripherals (`1` instead of `n`).
 //!
-//! ### Device driver compatible only with 7-bit addresses
+//! This API supports bus sharing natively. Types implementing [`I2c`] are allowed
+//! to represent either exclusive or shared access to an I2C bus. HALs typically
+//! provide exclusive access implementations. Drivers shouldn't care which
+//! kind they receive, they just do transactions on it and let the
+//! underlying implementation share or not.
+//!
+//! The [`embedded-hal-bus`](https://docs.rs/embedded-hal-bus) crate provides several
+//! implementations for sharing I2C buses. You can use them to take an exclusive instance
+//! you've received from a HAL and "split" it into mulitple shared ones, to instantiate
+//! several drivers on the same bus.
+//!
+//! # For driver authors
+//!
+//! Drivers can be generic over `I2c<SevenBitAddress>` or `I2c<TenBitAddress>` depending
+//! on the kind of address supported by the target device. If it can use either, the driver can
```

Comment:
```suggestion
//! Drivers can select the adequate address length with `I2c<SevenBitAddress>` or `I2c<TenBitAddress>` depending
//! on the target device. If it can use either, the driver can
```